### ISSUE TITLE
Gantt chart

### ISSUE THREAD

**Issue by simahao** [wekan/wekan#2870]
## This is not a issue

I want to know is there any plan for supporting gantta chart?

**Comment by xet7** [comment]
I can develop Gantt chart feature, if someone pre-pays me 2000 euro at:
https://wekan.team/commercial-support/

**Comment by xet7** [comment]
It seems from Wekan Survey 2020-01 results, and some questions that I get at email, that many would like to have Gantt chart feature.

Anyone can participate in funding Gantt chart feature by paying some part of 2000 euro at:
https://wekan.team/commercial-support/

I do develop Wekan full time, and pay my monthly bills like rent with income from Wekan Commercial Support I provide.

Alternatively, if some other Wekan contributor has time to develop Gantt chart feature, pull requests welcome.

**Comment by schulz** [comment]
I have integrated this gantt app into our customized Wekan: https://dhtmlx.com/docs/products/dhtmlxGantt/.

**Comment by xet7** [comment]
@schulz 

That dhtmlxGantt is available only with GPL or commercial license:
https://dhtmlx.com/docs/products/licenses.shtml

There is only benefit to Wekan if you send pull request with code and components that uses MIT or compatible license. Currently Wekan has most code with 89% MIT and BSD-3-clause licensed code. There is 11% Apache 2.0 licensed code, that I hope to be able remove sometime, because Apache and GPL are not copyfree licenses:
http://copyfree.org/standard/rejected

Wekan uses MongoDB, that has changed license from AGPL to SSPL. I wish sometime to find license compatible alternatives to MongoDB, or add support for other databases.

For an example about MIT-licensed Gantt code, you can look at this plugin:
https://github.com/kanboard/plugin-gantt

Also see:
https://blog.wekan.team/2018/02/benefits-of-contributing-your-features-to-upstream-wekan/

Wekan is currently free also for commercial use, and Wekan is also integrated to many other projects. All of Wekan is Open Source, there is not currently anything propietary. I would presume making Wekan GPL would also affect those other projects.

Or, would someone be really OK about Wekan changing license to GPL ?

**Comment by ocdtrekkie** [comment]
I think your commercial users would rather you not GPL. But people might be interested in using this fork if they want this feature.

**Comment by xet7** [comment]
Well, if @schulz could publish code at GitHub of somewhere else publicly, with code as MIT license and Gantt component could be changed to MIT-licensed one, that would work too.

**Comment by xet7** [comment]
Some have asked me that "is there Enterprise version of Wekan", and currently the Wekan Open Source version is full Enterprise version with all the same features. There is no separate install of Enterprise version.

**Comment by xet7** [comment]
Wekan is Open Source because:
- Wekan has been Open Source from start about 5.5 years ago by original creator of Wekan. Then about 3.5 years I continued maintaining Wekan, first as a hobby for free, and after that as community request partly funded by bounties and my [Wekan Team Commercial Support](https://wekan.team/commercial-support/). Upcoming prepaid features and fixes are listed at my [Roadmap](https://boards.wekan.team/b/D2SzJKZDS4Z48yeQH/wekan-open-source-kanban-board-with-mit-license).
- Wekan global community sends [translations](https://transifex.com/wekan/wekan) and [pull requests](https://github.com/wekan/wekan/pulls) to Wekan for features and fixes. Wekan is community driven project. Wekan has about total 180 contributors during about 5.5 years of Wekan development. Combined total amount of time used for coding Wekan is 13 years using COMOCO model.

**Comment by schulz** [comment]
> Well, if @schulz could publish code at GitHub of somewhere else publicly, with code as MIT license and Gantt component could be changed to MIT-licensed one, that would work too.

So at the moment I can't publish Wekan + dhtmlxGantt on my github account? Or can I?

**Comment by xet7** [comment]
@schulz 

Of course you can publish it. You just need to remove old fork of Wekan from your GitHub account, and then add your newest Wekan + dhtmlxGantt repo there.

**Comment by xet7** [comment]
All those licenses allow to use, publish and redistribute code.

**Comment by xet7** [comment]
With your code, I can create up-to-date Wekan version with dhtmlxGantt, and make separate Docker image for it, for anyone to try.

dhtmlxGantt beight GPL does not impact normal Wekan users. Anyone is free to use and redistribute original and modified GPL code.

Only ones GPL affects is those that would like to very deeply integrate Wekan to other software and they would not like to publish their own source code. For those, sometime later I will try to look can dhtmlxGantt component be changed to some other similar MIT licensed component.

Thanks a lot for your contribution!

**Comment by xet7** [comment]
@schulz

Do you have time to publish your code at your GitHub account? It would help all Wekan users around the world, Wekan community would be very grateful.

**Comment by schulz** [comment]
@xet7 sure, will do it. Our version of Wekan is very customized, I need to make a fresh fork and then integrate the gantt chart. Will take me a couple of days, I'll post the updates here.

**Comment by xet7** [comment]
@schulz 

Thanks a lot ! Having changes in a fresh fork and new pull request makes it a lot easier to integrate it into Wekan.

**Comment by schulz** [comment]
@xet7 I have finished the integration: https://github.com/schulz/wekan. I've taken the freshest versions of Wekan and DhtmlxGantt.

**Comment by xet7** [comment]
@schulz 

Thanks!

BTW, today released Wekan v4.40 has custom logo settings in Admin Panel/Layout.
https://github.com/wekan/wekan/wiki/Custom-Logo

**Comment by ocdtrekkie** [comment]
I do think the explicit statement "To use dhtmlxGantt in non-GPL projects (and get Pro version of the product), please obtain Commercial/Enterprise or Ultimate license on our site" means it'd be hard to integrate this, formally, even as just a package added to Wekan. But perhaps it might be worth asking the company how they would feel about it?

**Comment by ocdtrekkie** [comment]
Alternately, if you wanted to allow people who were willing to get licenses for dhtmlxGantt to use it with Wekan without forking, you could maybe disable it until someone put in a license key from dhtmlxGantt or whatever? I don't know. Maybe ask them.

**Comment by xet7** [comment]
@ocdtrekkie 

I have plans to change that Gantt component to MIT-licensed Gantt component.

**Comment by xet7** [comment]
Those Gantt changes are small amount of code, it looks very clean.

**Comment by xet7** [comment]
Well, depends when someone has time to look at [MIT-licensed component like this](https://github.com/kanboard/plugin-gantt) and submits pull request to Wekan.

Anyway, I happily created Wekan GPLv2 Gantt version that will be here, when builds finish:
- https://github.com/wekan/wekan-gantt-gpl
- https://snapcraft.io/wekan-gantt-gpl
- https://hub.docker.com/repository/docker/wekanteam/wekan-gantt-gpl
- https://quay.io/wekan/wekan-gantt-gpl

I think I will now concentrate to coding #2934 and #3173 that are urgent.

**Comment by xet7** [comment]
Those Gantt changes are here https://github.com/wekan/wekan-gantt-gpl/commit/2fdaf425536bad7b6eae6f03cd7bd1db1e40f9da

**Comment by avatar1024** [comment]
@xet7 

Hello,

Firstly thanks a lot for packing this and creating the gpl version! 

As far as I understand from this discussion, it's pretty unlikely this plugin will ever be merged with the original version, as it would impose the GPL restrictions (which personally I thought could be interesting as it could mean some commercial users would (have to) release some of their code in the open to keep using Wekan, and this might help other to deeply integrate the app - but I can see it might be a risky thing to do to upset a range of customers!). It therefore feels likely this gpl version will get dropped if someone bothers creating a plugin compatible with current MIT license.

The question therefore is: is there compatibly risk in to switch to the gpl version of the app now (as an update) and then switch back to the non gpl later (also as an update)?

**Comment by xet7** [comment]
@avatar1024 

Those changes to/from Gantt GPL version are very small. That Gant Chart only displays some data exact same Wekan database on both MIT and GPL versions, there is not even any differences in database structure.

For newest Wekan changes, I have released all newest Wekan versions for all of 1) Wekan 2) Wekan Gantt GPL 3) Wekan Ondra versions. Wekan Ondra is from time before creating official Wekan Snap, only difference there is snap name. If sometime for example 2) and 3) would deprecated so that newest update shows only migration instructions at Wekan webpage, then those users should do mongodump/mongorestore to newest Wekan snap, like this:

## Changing from Wekan to Wekan Gantt GPL

1) Install newest MongoDB to have also mongorestore available

2) Backup database and settings:
```
sudo snap stop wekan.wekan

mongodump --port 27019

snap get wekan > snap-set.sh

sudo snap remove wekan

sudo snap install wekan-gantt-gpl

sudo snap stop wekan-gantt-gpl.wekan

nano snap-set.sh
```
Then edit that textfile so all commands will be similar to this:
```
sudo snap set wekan-gantt-gpl root-url='https://example.com'
```
And run settings:
```
chmod +x snap-set.sh

./snap-set.sh

sudo snap start wekan-gantt-gpl.wekan
```
## Changing from Wekan Gantt GPL to Wekan

1) Install newest MongoDB to have also mongorestore available

2) Backup database and settings:
```
sudo snap stop wekan-gantt-gpl.wekan

mongodump --port 27019

snap get wekan-gantt-gpl > snap-set.sh

sudo snap remove wekan-gantt-gpl

sudo snap install wekan

sudo snap stop wekan.wekan

nano snap-set.sh
```
Then edit that textfile so all commands will be similar to this:
```
sudo snap set wekan root-url='https://example.com'
```
And run settings:
```
chmod +x snap-set.sh

./snap-set.sh

sudo snap start wekan.wekan
```

**Comment by xet7** [comment]
I just got this feedback from someone using that Wekan Gantt GPL version:

> Thanks, I can see Gantt menu item and I can click on it but none of the cards show up, even not those with start and due dates...

So does someone know how that Gantt feature works? Or does it work at all?

**Comment by avatar1024** [comment]
@xet7 Thanks a lot for the explanation and step by step guidance! 

And indeed, be good to know if the function is actually working ;) If no one responds I'll try to install that version in another environment and report back.

**Comment by schulz** [comment]
> I just got this feedback from someone using that Wekan Gantt GPL version:
> 
> > Thanks, I can see Gantt menu item and I can click on it but none of the cards show up, even not those with start and due dates...
> 
> So does someone know how that Gantt feature works? Or does it work at all?

Both Start and End dates should be set (not Due date) for the tasks to be displayed.

**Comment by xet7** [comment]
@schulz 

In Gantt chart, is there possibility to zoom out to weeks or months view?

If not, can you add it with pull request to this repo?
https://github.com/wekan/wekan-gantt-gpl

**Comment by schulz** [comment]
> @schulz
> 
> In Gantt chart, is there possibility to zoom out to weeks or months view?
> 
> If not, can you add it with pull request to this repo?
> https://github.com/wekan/wekan-gantt-gpl

Will check it.

**Comment by xet7** [comment]
Thanks!

**Comment by xet7** [comment]
As a note to myself, Gantt related changes:
- Original changes, from schulz https://github.com/wekan/wekan-gantt-gpl/commit/2fdaf425536bad7b6eae6f03cd7bd1db1e40f9da
- Make some parts of Gantt chart translateable and fix lint errors, from xet7
  - https://github.com/wekan/wekan-gantt-gpl/commit/52bbff65ac4c03193bcddd78c9544a155d0ba62c
  - https://github.com/wekan/wekan-gantt-gpl/commit/fd363c69cc6e1cf3a283e3dbcc323edb1eae896e
  - https://github.com/wekan/wekan-gantt-gpl/commit/38bcb8a72416fc4a91bf5fb72455480c18d374ff

**Comment by chrisi51** [comment]
are there any plans to integrate gantt to the regular wekan version and especially make it available in univention ucs on that way?

**Comment by xet7** [comment]
@chrisi51 

> @xet7 are there any plans to integrate gantt to the regular wekan version and especially make it available in univention ucs on that way?

**Summary**: Because nobody has prepaid me for additional development of this feature, it's easy for me to find excuses to look at this maybe sometime later.

Also depends is current Gantt version useful in any way. If I would publish it directly to Univention, then in future there should be  compatible migration to MIT licensed version.

Originally in this issue at 2nd comment https://github.com/wekan/wekan/issues/2870#issuecomment-570903539 I asked 2000 euro to develop this feature, nobody funded that. So after that, @schulz contributed code that used GPL licensed Gantt component. At https://github.com/wekan/wekan/issues/2870#issuecomment-721725421 I asked @schulz about possibility to zoom out to weeks or months view, no reply yet.

About future development plans, one of my customers asked about more Gantt features:

> ok that works Tks but how can i zoom out to weeks or months view??
> 
> Any news on having the gantt chart working properly:
> - zoom in and out
> - task dependency links. Task dependency means that you can drag and drop e link between task x and task y in gantt chart so that the relation between the 2 is set.. Normally this means that task x has to be finished before task y can start etc. etc..
>
> I agree a MIT licensed Gantt chart would be better. 
>
> Normal Gantt chart works like below:
>
> - Tasks are viewed per start-end date
> - you can drag a dependency line from eg. end of task x to beginning of tasks y
> - if done so the start date of task y is automatically shifted to end date of task x
> - Normally you can also create subtasks and then the main tasks dates are determined by the individual dates of subtasks
> - etc etc.
> - follow this link about gantt charts: https://www.gantt.com
>
> How much would it costs to have it look properly?

I replied:
> I would prefer to finish Teams/Organizations and some other features before looking at adding more to Gantt feature. Would that be OK ?

Customer replied:
> Ok no problem..

**Comment by schulz** [comment]
@xet7 haven't chance to look into a zoom in/zoom out feature, I'm quite busy with my main jobs. But I still plan to look into it when I'll work on my version of Wekan.
I have some more features:
- adding comments by replying via email
- 2FA authentication

Can add these as well but need to find time.

**Comment by diegodorgam** [comment]
Seems like a very promising feature, congratulations on the initiative guys!
Looking forward to have it implemented. 
Would make some things different though. Seems like you guys have invested on it for some time, and I would like to do it too, so please don't mind me asking a few questions, and please don't bother to answer if you think those are just nonsense:

1. Gantt it's a formidable way to plan time tables along the project, beyond classical agile boards, sprints and etc. So, would be possible to have the tasks created without dates (init, due and close) to be listed in the gantt and have their dates defined in an interactive way, like by clicking in the chart? (kinda like OpenProject does)
2. It's also a great tool to create the project WBS, but in order to do that you need to be able to add tasks and to ident then inside other tasks (transforming the parent task into a work package), which I think it not necessarily should transform the child task into a sub task, or at least not a sub board task. You guys think would be possible to have a user friendly component which enables to create WBS and translate than into the board, with columns translated as task states ?

I've been working with wekan for the last two years and now, my company has grown and I do miss some interactive gantt and portfolio management features, so I've started searching for other solutions like OpenProject, but honestly, I would not like to change it, we're kinda in love with wekan already.

So if there's anyway we could improve this feature, help to put it out sooner, please let me know.
Thanks for your attention.

**Comment by xet7** [comment]
@diegodorgam 

I don't know how WBS or OpenProject works, you could add:
- recorded animgifs about how WBS and OpenProject works
- screenshot step-by-step mockups about how it should work. Usually from many different options best looking one is selected, and MIT or BSD licensed components preferred.

There is kind of many overlapping feature requests:
- For calendar, currently Wekan uses https://fullcalendar.io/ , but it's Gantt component is propietary. There are some forks of fullcalendar that have MIT licensed Gantt code, but those forks usually are not up-to-date.
- There is some planning for [Timesheet / Time Tracking with Kimai](https://github.com/wekan/wekan/issues/812). I have not looked does Kimai have some Gantt related etc reporting. That Kimai integration probably gets funding sometime in near future.
- There was originally some thinking about [changing Summernote wysiwyg editor so something else](https://github.com/wekan/wekan/issues/3399), but it got reverted because Summernote mostly works already, and it's only needed to fix some related bugs. Anyway, in that issue is comment about MIT licensed [Toast UI](https://github.com/wekan/wekan/issues/3399#issuecomment-751915295) and other [Toast UI components](https://github.com/nhn) like [Toast UI Calendar](https://github.com/nhn/tui.calendar). I have not looked does Toast UI Calendar have Gantt chart.
- At [npmjs.com](https://npmjs.com) is also some Gantt components with MIT license.
- There is feature request for [Multi Board Calendar like Planyway from Trello](https://github.com/wekan/wekan/issues/2469), but it is not funded yet.
- There is feature request for [Task Dependencies](https://github.com/wekan/wekan/issues/3210)

In general:
- It would be nice if with only one calendar component all related features could be implemented. Like fixing existing fullcalendar to also have Gantt, etc.
- Fixing existing components is less work than adding new component, or changing component.

**Comment by xet7** [comment]
Fastest way to get some feature or fix to Wekan, is look at source code, figure it out, and send pull request. Many Wekan contributors have just somehow figured it out and sent features and fixes.

**Comment by benjaminhrivera** [comment]
@schulz Hey! Are there any other features from the Gantt NPM that you've worked on? I was able to retain links created between tasks, as before they were not being saved

**Comment by xet7** [comment]
@benjaminhrivera 

How did you retain links between tasks?

**Comment by xet7** [comment]
@benjaminhrivera 

Screenshot?

**Comment by benjaminhrivera** [comment]
We started by adding new items to the cards.js schema
![image](https://user-images.githubusercontent.com/57966706/121256796-f06a5d80-c87a-11eb-9621-db6bd7c953be.png)

then added these 2 functions to the cards.helper function in cards.js
![image](https://user-images.githubusercontent.com/57966706/121257244-7686a400-c87b-11eb-8545-5e6884f60a87.png)

adding this to gantt.js which will fire during the events specified
![image](https://user-images.githubusercontent.com/57966706/121256592-b731ed80-c87a-11eb-9941-44603728ab67.png)

and lastly parsing the information the same way the wekan cards are being passed in
![image](https://user-images.githubusercontent.com/57966706/121257703-02003500-c87c-11eb-956b-bbcb7e71043a.png)

@xet7 hope that helps

**Comment by benjaminhrivera** [comment]
_**First image**_

 targetId_gantt: {
      /**
       * ID of card which is the child link in gantt view
       */
      type: [String],
      optional: true,
      defaultValue: [],
    }, 
    linkType_gantt: {
      /**
       * ID of card which is the parent link in gantt view
       */
      type: [Number],
      decimal: false,
      optional: true,
      defaultValue: [],
    },
    linkId_gantt: {
      /**
       * ID of card which is the parent link in gantt view
       */
      type: [String],
      optional: true,
      defaultValue: [],
    }, 

_**second image**_

setGanttTargetId(sourceId, targetId, linkType, linkId){
    return Cards.update({ _id: sourceId}, {
      $push: {
        targetId_gantt: targetId,
        linkType_gantt : linkType,
        linkId_gantt: linkId
      }
    });
  },
  removeGanttTargetId(sourceId, targetId, linkType, linkId){
    return Cards.update({ _id: sourceId}, {
      $pull: {
        targetId_gantt: targetId,
        linkType_gantt : linkType,
        linkId_gantt: linkId
      }
    });
  }

**_third image_**

gantt.attachEvent("onLinkCreated", function(link){
      const c = Cards.findOne(link.source);
      var sourceTask = gantt.getTask(link.source);
      var targetTask = gantt.getTask(link.target);

      c.setGanttTargetId(sourceTask.id, targetTask.id, link.type, link.id);
      return true;
    });

    gantt.attachEvent("onBeforeLinkDelete", function(id,item){
      const c = Cards.findOne(item.source);

      c.removeGanttTargetId(item.source, item.target, item.type, id)
      return true;
    });

**_last image_**

this.autorun(() => {
      const events = [];
      const LinksData = [];
      currentBoard.cards().forEach(c => {
        if (!c.startAt || !c.endAt) {
          return;
        }

        const event = {
          id: c._id,
          text: c.title,
          description: c.description,
          start_date: moment(c.startAt).format('DD-MM-YYYY'),
          due_date: moment(c.endAt).format('YYYY-MM-DD'),
          duration: moment.duration(moment(c.endAt).diff(c.startAt)).asDays(),
          //type: projectTask,
          parent: c.parentId,
        };

        if(c.targetId_gantt) {
          for (var i = 0; i < c.targetId_gantt.length; i++) {
            var cardLink = c.targetId_gantt[i];
            var linkType = c.linkType_gantt[i];
            var linkId = c.linkId_gantt[i];

            const LinkInfo= {
              id: linkId,
              source: c._id,
              target: cardLink,
              type: linkType,
            };
          LinksData.push(LinkInfo);
          }
        }

        events.push(event);
        //LinksData.push(links);
      });

      //this is the reason why links get deleted when a task is moved... why is this here?
      gantt.clearAll();

      gantt.parse({
        data: events,
        links: LinksData,
      });
    });
  },

**Comment by schulz** [comment]
> @schulz Hey! Are there any other features from the Gantt NPM that you've worked on? I was able to retain links created between tasks, as before they were not being saved

Hey, haven't chance to work on wekan at all. I have some features but I need some time to make a PR on the latest wekan code:
- parsing email responses and adding them as comments
- 2-factor authentication

**Comment by chrisi51** [comment]
@xet7 just installed new UCS 5.71 and set start and end on one card in a test board. 

where should i see a gantt chart now? or do i have to activate this feature in any way? didnt found any setting in backend or board or card settings

**Comment by xet7** [comment]
@chrisi51 

I have not tested WeKan 5.71 yet at UCS yet, yes it has been released already. After setting dates like mentioned at https://github.com/wekan/wekan/wiki/Gantt , from board top bar click Swimlanes/Lists/Calendar/Gantt dropdown menu.

**Comment by chrisi51** [comment]
@xet7 ok i've found it and it seem to work :)

**Comment by xet7** [comment]
I upgraded WeKan Gantt GPLv2 to newest WeKan, Meteor, etc, and added translations at https://github.com/wekan/wekan-gantt-gpl/commit/a08c01c76eaaf2884890a39c97d1f72da222fba1 so that there was no Gantt related errors anymore at browser inspect console, but still Gantt chart is not visible.

I don't know is something missing still, like moment/translations related, or something else.

Related https://github.com/wekan/wekan/wiki/Gantt

**Comment by xet7** [comment]
https://quay.io/wekan/wekan-gantt-gpl

https://hub.docker.com/r/wekanteam/wekan-gantt-gpl/

**Comment by xet7** [comment]
Yes, Gantt feature is broken on WeKan Gantt GPL version. I presume that I (or someone else) will make MIT licensed Gantt feature to WeKan MIT license version sometime in the future. GPL version will be discontinued.

**Comment by xet7** [comment]
3rd  alternative is try to use Mermaid that is included in WeKan:
- https://mermaid-js.github.io
- https://github.com/wekan/wekan/wiki/Mermaid-Diagram

But I don't know does Mermaid have as many features as https://github.com/frappe/gantt

**Comment by xet7** [comment]
Mermaid was removed at [WeKan v6.76](https://github.com/wekan/wekan/blob/master/CHANGELOG.md#v676-2023-02-22-wekan--release) because it did not work anymore at all, it had many errors.

I like to keep WeKan MIT licensed, and not add any GPL Gantt component. Some MIT licensed Gantt is at above links in this issue.

It would make most sense to add Gant back after these In Progress huge amount of upgraded https://github.com/wekan/wekan/issues/4578#issuecomment-1511879386

**Comment by avatar1024** [comment]
Same that Mermaid didn't work well. But that Frappe Gantt looks alright!

**Comment by shalvy** [comment]
> b

was the zoom resolved

**Comment by xet7** [comment]
@shalvy 

What zoom?

**Comment by tromlet** [comment]
Do we have any news on this MIT-licensed Gantt chart functionality?

**Comment by xet7** [comment]
@tromlet 

Currently there is in progress adding other major upgrades, features and fixes, so I'm not yet sure will this feature come at this or next year.
